cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowHttpHandlerTest.java

```diff
@@ -1197,4 +1198,84 @@ private String appendScopeAndKeyToUrl(String workflowTokenUrl, @Nullable Workflo
+    waitState(workflowId, ProgramRunStatus.RUNNING.name());
+    waitState(workflowId, "STOPPED");
+
+    List<RunRecord> programRuns = getProgramRuns(workflowId, ProgramRunStatus.FAILED.name());
+    Assert.assertEquals(1, programRuns.size());
+
+    // Start program with "put.in.map" argument.
+    // It will perform put operation on the WorkflowToken in the map method of the Mapper class.
+    // This should fail.
+    outputPath = new File(tmpFolder.newFolder(), "output").getAbsolutePath();
+    startProgram(workflowId, ImmutableMap.of("inputPath", createInputForRecordVerification("secondInput"),
+                                             "outputPath", outputPath, "put.in.map", "true"));
+    waitState(workflowId, ProgramRunStatus.RUNNING.name());
+    waitState(workflowId, "STOPPED");
+
+    programRuns = getProgramRuns(workflowId, ProgramRunStatus.FAILED.name());
+    Assert.assertEquals(2, programRuns.size());
+
+    // Start program with "put.in.reducer.initialize" argument.
+    // It will perform put operation on the WorkflowToken in the Initialize method of the Reducer class.
+    // This should fail.
+    outputPath = new File(tmpFolder.newFolder(), "output").getAbsolutePath();
+    startProgram(workflowId, ImmutableMap.of("inputPath", createInputForRecordVerification("thirdInput"),
+                                             "outputPath", outputPath, "put.in.reducer.initialize", "true"));
+    waitState(workflowId, ProgramRunStatus.RUNNING.name());
+    waitState(workflowId, "STOPPED");
+
+    programRuns = getProgramRuns(workflowId, ProgramRunStatus.FAILED.name());
+    Assert.assertEquals(3, programRuns.size());
+
+    // Start program with "put.in.reduce" argument.
+    // It will perform put operation on the WorkflowToken in the reduce method of the Reducer class.
+    // This should fail.
+    outputPath = new File(tmpFolder.newFolder(), "output").getAbsolutePath();
+    startProgram(workflowId, ImmutableMap.of("inputPath", createInputForRecordVerification("fourthInput"),
+                                             "outputPath", outputPath, "put.in.reduce", "true"));
+    waitState(workflowId, ProgramRunStatus.RUNNING.name());
+    waitState(workflowId, "STOPPED");
+
+    programRuns = getProgramRuns(workflowId, ProgramRunStatus.FAILED.name());
+    Assert.assertEquals(4, programRuns.size());
+
+    // Start program with only inputPath and outputPath arguments.
+    // This should succeed.
+    outputPath = new File(tmpFolder.newFolder(), "output").getAbsolutePath();
+    startProgram(workflowId, ImmutableMap.of("inputPath", createInputForRecordVerification("fifthInput"),
+                                             "outputPath", outputPath));
+    waitState(workflowId, ProgramRunStatus.RUNNING.name());
+    waitState(workflowId, "STOPPED");
+
+    programRuns = getProgramRuns(workflowId, ProgramRunStatus.COMPLETED.name());
+    Assert.assertEquals(1, programRuns.size());
+  }
```

Comment:
Yes. We allow putting from beforeSubmit() and onFinish(). Added test case for the put operation from onFinish method in the new commit. We are already performing put operation in the beforeSubmit method - https://github.com/caskdata/cdap/pull/3339/files#diff-6d5c552cd9d1a2ee73692a5f4aa9c8f0R91